internal/tokenrefresher/tokenrefresher.go

```diff
@@ -0,0 +1,170 @@
+)
+
+// tokenReponse struct required to unmarshal the response from a POST token refresh request
+type tokenResponse struct {
+	AccessToken           string `json:"access_token"`
+	Type                  string `json:"token_type"`
+	ExpiresIn             int64  `json:"expires_in"`
+	RefreshToken          string `json:"refresh_token"`
+	RefreshTokenExpiresIn int64  `json:"refresh_expires_in"`
+	Scope                 string `json:"scope"`
+	CreatedAt             int64  `json:"created_at"`
+}
+
+func (t tokenResponse) String() string {
+	return fmt.Sprintf(
+		"CreatedAt: %v, Type: %v, ExpiresIn: %v, RefreshTokenExpiresIn: %v",
+		t.CreatedAt,
+		t.Type,
+		t.ExpiresIn,
+		t.RefreshTokenExpiresIn,
+	)
+}
+
+// RefresherTokenStore is an interface used for refreshing tokens stored by the gateway
+type RefresherTokenStore interface {
+	GetRefreshToken(context.Context, string) (models.OauthToken, error)
+	GetAccessToken(context.Context, string) (models.OauthToken, error)
+	SetRefreshToken(context.Context, models.OauthToken) error
+	SetAccessToken(context.Context, models.OauthToken) error
+	GetExpiringAccessTokenIDs(context.Context, time.Time, time.Time) ([]string, error)
+}
+
+// ScheduleRefreshExpiringTokens intialises a gocron job to run refreshExpiringTokens at a specified interval
+func ScheduleRefreshExpiringTokens(
+	ctx context.Context,
+	tokenStore RefresherTokenStore,
+	gitlabClientID string,
+	gitlabClientSecret string,
+	minsToExpiration int,
+) error {
+	s := gocron.NewScheduler(time.UTC)
+	job, err := s.Every(minsToExpiration).
+		Minutes().
+		Do(refreshExpiringTokens, ctx, tokenStore, gitlabClientID, gitlabClientSecret, minsToExpiration)
+	s.StartBlocking()
+	if err != nil {
+		slog.Error("Starting gocron job failed", "error", err)
+	} else {
+		slog.Info("Job starting", "job", job)
+	}
+	return err
+}
+
+// refreshExpiringTokens refreshes tokens in the token store expiring in the next minsToExpiration minutes
+func refreshExpiringTokens(
+	ctx context.Context,
+	tokenStore RefresherTokenStore,
+	clientID string,
+	clientSecret string,
+	minsToExpiration int,
+) error {
+	// Get a list of expiring access tokens ids in the next minsToExpiration minutes
+	expiringTokenIDs, err := tokenStore.GetExpiringAccessTokenIDs(
+		ctx,
+		time.Now(),
+		time.Now().Add(time.Minute*time.Duration(minsToExpiration)),
+	)
+	if err != nil {
+		slog.Error("GetExpiringAccessTokenIDs failed", "error", err)
+		return err
+	}
+
+	// For each token id expiring in the next minsToExpiration minutes
+	for _, expiringTokenID := range expiringTokenIDs {
+
+		// Get the refresh and access tokens associated with the token ID
+		myRefreshToken, err := tokenStore.GetRefreshToken(ctx, expiringTokenID)
+		if err != nil {
+			slog.Error("GetRefreshToken failed", "error", err)
+			return err
+		}
+
+		myAccessToken, err := tokenStore.GetAccessToken(ctx, expiringTokenID)
+		if err != nil {
+			slog.Error("GetAccessToken failed", "error", err)
+			return err
+		}
```

Comment:
should we collect errors here and return them all at once later? otherwise one error early on prevents all other tokens from being refreshed

Reply:
Yes we def should. Done.